### ISSUE TITLE
Can't set shell && windows-shell || windows-powershell if OS == Windows

### ISSUE THREAD

**Issue by runeimp** [casey/just#1293]
In the current version of Just (v1.3.0) I can do something like

```
set windows-powershell := true
set shell := ["pwsh", "-c"]
```

-or-

```
set windows-shell := ["powershell", "-Command"]
set shell := ["pwsh", "-c"]
```

and be able to use PowerShell scripting in recipes on macOS Big Sur with PowerShell Core v7.2.5 installed no problem. But running the same Justfile in a Windows 10 VirtualBox VM errors with `error: Recipe `_term-wipe` could not be run because just could not find the shell: program not found`. In this instance the `_term-wipe` recipe simply has a single line `Clear-Host`. If I comment out `set shell := ["pwsh", "-c"]` it works fine in Windows 10.

I was really hoping to finally having a good cross-platform solution.  😞 (guilt, guilt, GUILT!)

Just kidding, hoping this is an easy fix to resolve. I don't mind installing PowerShell Core on the only Mac doing development in the company. But was really hoping not to have to install PowerShell Core on all the VMs and user systems that already have Windows 10 and PowerShell 5.1 already installed.  👼

**Comment by casey** [comment]
Hmm, not super familiar with how Windows finds binaries. Could it be that you need:

```
set windows-shell := ["powershell.exe", "-Command"]
set shell := ["pwsh", "-c"]
```

Or a full path?

No guilt, I just want it to work too!

**Comment by runeimp** [comment]
Everything works fine without specifying the `.exe` as long as both settings are not active at the same time.

It works fine the second I remove `set shell := ["pwsh", "-c"]`.

The expectation is that if a `set windows-*` is active it will override anything in `set shell`. On macOS I believe there is no problem because anything that starts out `set windows-*` is, most likely, completely ignored or non-existent in the Mac build. But on Windows `set shell` takes precedence for some reason or there is some general conflict.

**Comment by casey** [comment]
Found the issue! When I originally implemented the feature I missed testing different combinations of shell settings, and completely bungled it. This should be an easy fix, but I'm currently pretty slammed, so won't have time to fix it until Saturday.

**Comment by runeimp** [comment]
Oh, the anguish... two... whole... days...

No prob. Glad it's an easy fix.  😉

**Comment by casey** [comment]
I got to it a little early. Is it convenient for you to build from source? If so, could you try this branch: https://github.com/casey/just/pull/1297 The fix is pretty simple, I wrote tests that pass, and it should work, but I'd like to make sure before I merge it.

**Comment by runeimp** [comment]
Hm, I install Rust and test tomorrow.  😁

**Comment by runeimp** [comment]
Sorry, got busy yesterday and most of today with the family.  👼 

```sh
$ git clone https://github.com/casey/just
$ cd just
$ git fetch --all
$ git checkout test-windows-shell
$ just
cargo test
   Compiling libc v0.2.126
   Compiling proc-macro2 v1.0.41
   Compiling quote v1.0.20
   Compiling syn v1.0.98
   Compiling memchr v2.5.0
   Compiling typenum v1.15.0
   Compiling rustversion v1.0.8
   Compiling serde_derive v1.0.140
   Compiling doc-comment v0.3.3
   Compiling log v0.4.17
   Compiling serde v1.0.140
   Compiling remove_dir_all v0.5.3
   Compiling serde_json v1.0.82
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-m64" "-arch" "x86_64" "/var/folders/9x/p3j02cpn1vg5317gwv_r73vw0000gn/T/rustcek8dDw/symbols.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.0.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.1.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.10.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.11.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.12.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.13.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.14.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.15.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.2.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.3.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.4.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.5.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.6.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.7.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.8.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.9.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.5gaaxcv68al5j5d4.rcgu.o" "-L" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/deps" "-L" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libstd-b70ac28315bc2a94.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libpanic_unwind-5e9faae75d9b1d88.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libobject-ca0da9b068fe6d21.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libmemchr-9930c301c0009607.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libaddr2line-7a788fe0d6905a91.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libgimli-23f839bbe5759c90.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_demangle-c59a614762cb8623.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libstd_detect-13e083ce872067ef.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libhashbrown-55d2ec67aca15dfc.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libminiz_oxide-1156efb2cbf3bcf2.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libadler-6a1369f362a5d571.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_alloc-5cc492ea38a64c46.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libunwind-91fb2b2444b8858d.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libcfg_if-084ff94c77ce12d3.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/liblibc-4a6afe107171d911.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/liballoc-57aee859695ae695.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_core-a2ac79b335d55ca9.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libcore-12d21acdbb192066.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-91ed0700e4a5eb2a.rlib" "-lSystem" "-lresolv" "-lc" "-lm" "-liconv" "-L" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib" "-o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac" "-Wl,-dead_strip" "-nodefaultlibs"
  = note: ld: library not found for -lSystem
          clang: error: linker command failed with exit code 1 (use -v to see invocation)

$ justfile build
cargo build
   Compiling libc v0.2.126
   Compiling proc-macro2 v1.0.41
   Compiling quote v1.0.20
   Compiling syn v1.0.98
   Compiling memchr v2.5.0
   Compiling typenum v1.15.0
   Compiling rustversion v1.0.8
   Compiling serde_derive v1.0.140
   Compiling log v0.4.17
   Compiling serde v1.0.140
   Compiling doc-comment v0.3.3
   Compiling unicode-width v0.1.9
   Compiling camino v1.0.9
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-m64" "-arch" "x86_64" "/var/folders/9x/p3j02cpn1vg5317gwv_r73vw0000gn/T/rustc5AYvm3/symbols.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.0.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.1.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.10.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.11.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.12.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.13.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.14.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.15.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.2.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.3.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.4.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.5.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.6.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.7.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.8.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.build_script_build.642b96c5-cgu.9.rcgu.o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac.5gaaxcv68al5j5d4.rcgu.o" "-L" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/deps" "-L" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libstd-b70ac28315bc2a94.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libpanic_unwind-5e9faae75d9b1d88.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libobject-ca0da9b068fe6d21.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libmemchr-9930c301c0009607.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libaddr2line-7a788fe0d6905a91.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libgimli-23f839bbe5759c90.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_demangle-c59a614762cb8623.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libstd_detect-13e083ce872067ef.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libhashbrown-55d2ec67aca15dfc.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libminiz_oxide-1156efb2cbf3bcf2.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libadler-6a1369f362a5d571.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_alloc-5cc492ea38a64c46.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libunwind-91fb2b2444b8858d.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libcfg_if-084ff94c77ce12d3.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/liblibc-4a6afe107171d911.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/liballoc-57aee859695ae695.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_core-a2ac79b335d55ca9.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libcore-12d21acdbb192066.rlib" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-91ed0700e4a5eb2a.rlib" "-lSystem" "-lresolv" "-lc" "-lm" "-liconv" "-L" "/usr/local/Cellar/rust/1.62.1/lib/rustlib/x86_64-apple-darwin/lib" "-o" "/Users/mgardner/Dropbox/Profile/Home/dev/lang/rust/just/target/debug/build/log-6b1ed19a4992e6ac/build_script_build-6b1ed19a4992e6ac" "-Wl,-dead_strip" "-nodefaultlibs"
  = note: ld: library not found for -lSystem
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Not sure what the problem is. I have Apple Command Line Tools installed.

**Comment by runeimp** [comment]
Should have all the latest C tools from Apple

```sh
$ which -a cc
/Library/Developer/CommandLineTools/usr/bin/cc
/usr/bin/cc
$ which -a ld
/Library/Developer/CommandLineTools/usr/bin/ld
/usr/bin/ld
```

**Comment by casey** [comment]
Hmmm, I don't recognize those errors. Actually, I just realized that my fix is buggy. I've got a bunch of stuff going, so I won't be able to get to it for a few days, so hold off for now.

**Comment by runeimp** [comment]
No problem. Just glad you're looking into it.  😁

**Comment by casey** [comment]
Okay, I finally got to this, and it was much trickier than I thought! I had basically totally forgotten how the different ways to configure the shell (command line options and settings) interacted, so I had to figure that out again.

I think I fixed it in this PR: https://github.com/casey/just/pull/1306

I pushed the code to a tag, which was built by CI, so you can download a binary to try it out here: https://github.com/casey/just/releases/tag/fix-windows-shell-setting-precedence-tag

I also documented the new behavior:

https://github.com/casey/just/pull/1306/commits/ae7dcf008833ad91db10f50a53693b7b6eb71b85

Let me know if it works and the description in the readme looks good!

**Comment by runeimp** [comment]
SUCCESS!!!

True Cross-Platform with PowerShell using the following settings/example:

### Cross-Platform with PowerShell Core on Windows and other platforms

```pwsh
pwsh_path := if os() == 'windows' { 'pwsh' } else { '/usr/bin/env pwsh' }
set shell := ["pwsh", "-NoLogo", "-Command"]

# PowerShell Testing
pwsh_shebang:
	#!{{pwsh_path}}
	$PSV = $PSVersionTable.PSVersion | % {"$_" -split "\." }
	$psver = $PSV[0] + "." + $PSV[1]
	if ($PSV[2].Length -lt 4) {
		$psver += "." + $PSV[2] + " Core"
	} else {
		$psver += " Desktop"
	}
	echo "PowerShell $psver"
```

### Cross-Platform with PowerShell Desktop on Windows and PowerShell Core on other platforms

```pwsh
pwsh_path := if os() == 'windows' { 'powershell' } else { '/usr/bin/env pwsh' }
set windows-shell := ["powershell", "-c"]
set shell := ["pwsh", "-NoLogo", "-Command"]

# PowerShell Testing
pwsh_shebang:
	#!{{pwsh_path}}
	$PSV = $PSVersionTable.PSVersion | % {"$_" -split "\." }
	$psver = $PSV[0] + "." + $PSV[1]
	if ($PSV[2].Length -lt 4) {
		$psver += "." + $PSV[2] + " Core"
	} else {
		$psver += " Desktop"
	}
	echo "PowerShell $psver"
```

**Comment by runeimp** [comment]
I recommend adding something like the above regarding the `README.md`. Also I saw that `windows-shell` is not in your _Table of Settings_.

**Comment by casey** [comment]
Nice!

I just updated the readme, and added [this example](https://github.com/casey/just/pull/1306/files#diff-c98f8a3e4a4a0650b135899e0132649c1c2e4a9879afae41da263be56eb15878).

Can you try it out and make sure that I got everything right? I used `windows-shell` only, because I didn't want to confuse people with multiple settings, but added a note about using `pwsh.exe` if it's available.

**Comment by runeimp** [comment]
Yep, you're good I think.  😁

**Comment by casey** [comment]
Great! Just released [1.4.0](https://github.com/casey/just/releases/tag/1.4.0), which includes the change. If Windows binaries aren't already up, then they should be soon. Thanks for sticking with this!
